<a href="https://colab.research.google.com/github/saida1997/ETE-456/blob/main/1608014_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Experiment Name: Exploring Different Types of Regression**
Course No: ETE 456

Name: Saida Sharmin

 ID: 1608014

**Objective**

> 1.To apply different regression algorithms in real world dataset.



**Task**

> Prediction of Insurance Costs



In [ ]:
import warnings
warnings.filterwarnings("ignore")

**Importing The Libraries**

In [ ]:
import numpy as np        
import pandas as pd     
import matplotlib.pyplot as plt 

**Dataset**

In [ ]:
# Download the data
!wget -O insurance.csv https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0

--2021-12-01 15:59:48--  https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.3.18, 2620:100:6018:18::a27d:312
Connecting to www.dropbox.com (www.dropbox.com)|162.125.3.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/mwgqgjbmfw0xa5p/insurance.csv [following]
--2021-12-01 15:59:48--  https://www.dropbox.com/s/raw/mwgqgjbmfw0xa5p/insurance.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uca9b04445215da91177f1e35e05.dl.dropboxusercontent.com/cd/0/inline/BbDtWH_GPU7EbRKdXQP5tBELTbGguckirHkgkJZq8-dyCwGSp3stt3OWYOzrugNB0D0k1HoOsAWNqknaTbdKWRnMaI0u7hl4n-TTpHz5EpOdgC46ijXebDb3jMeYWeH7ET2asIJNC7gAoMmjGc6_1oQj/file# [following]
--2021-12-01 15:59:48--  https://uca9b04445215da91177f1e35e05.dl.dropboxusercontent.com/cd/0/inline/BbDtWH_GPU7EbRKdXQP5tBELTbGguckirHkgkJZq8-dyCwGSp3stt3OWYOzrugNB0D0k1HoOsAWNqknaTb

**Importing the dataset**

In [ ]:


dataset = pd.read_csv('insurance.csv')
dataset

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

Now splitting the dataset into independent and dependent variables. Independent are the features which are used for prediction. Dependent is the target which will be predicted.

In [ ]:
features = dataset[['age'	,'sex',	'bmi','children','smoker','region']]  # for independent
target = dataset[['charges']]   # for dependent

In [ ]:
features

age     sex     bmi  children smoker     region
0      19  female  27.900         0    yes  southwest
1      18    male  33.770         1     no  southeast
2      28    male  33.000         3     no  southeast
3      33    male  22.705         0     no  northwest
4      32    male  28.880         0     no  northwest
...   ...     ...     ...       ...    ...        ...
1333   50    male  30.970         3     no  northwest
1334   18  female  31.920         0     no  northeast
1335   18  female  36.850         0     no  southeast
1336   21  female  25.800         0     no  southwest
1337   61  female  29.070         0    yes  northwest

[1338 rows x 6 columns]

In [ ]:
target

charges
0     16884.92400
1      1725.55230
2      4449.46200
3     21984.47061
4      3866.85520
...           ...
1333  10600.54830
1334   2205.98080
1335   1629.83350
1336   2007.94500
1337  29141.36030

[1338 rows x 1 columns]

**Taking Care of Missing Data**

> Here 'sklearn' means sckit learn which is a package used for machine learning model. It contains many libraries.

preprocessing is a library which contains several classes,methods. From this library we import the Imputer class which will allow us to take care of the missing data



In [ ]:
from sklearn.impute import SimpleImputer

In [ ]:
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean") # imputer is an object of Imputer class 
imputer = imputer.fit(features[['age','bmi','children']]) 

**fit()**: used for generating learning model parameters from training data

**transform()** : parameters generated from fit() method,applied upon model to generate transformed data set.

**fit_transform()** : combination of fit() and transform() api on same data set

Now we have to replace the missing value by mean into X

In [ ]:
features[['age','bmi','children']]= imputer.transform(features[['age','bmi','children']])

In [ ]:
features

age     sex     bmi  children smoker     region
0     19.0  female  27.900       0.0    yes  southwest
1     18.0    male  33.770       1.0     no  southeast
2     28.0    male  33.000       3.0     no  southeast
3     33.0    male  22.705       0.0     no  northwest
4     32.0    male  28.880       0.0     no  northwest
...    ...     ...     ...       ...    ...        ...
1333  50.0    male  30.970       3.0     no  northwest
1334  18.0  female  31.920       0.0     no  northeast
1335  18.0  female  36.850       0.0     no  southeast
1336  21.0  female  25.800       0.0     no  southwest
1337  61.0  female  29.070       0.0    yes  northwest

[1338 rows x 6 columns]

In [ ]:
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean") # imputer is an object of Imputer class 
imputer = imputer.fit(target[['charges']]) 

In [ ]:
target[['charges']]= imputer.transform(target[['charges']])

In [ ]:
target

charges
0     16884.92400
1      1725.55230
2      4449.46200
3     21984.47061
4      3866.85520
...           ...
1333  10600.54830
1334   2205.98080
1335   1629.83350
1336   2007.94500
1337  29141.36030

[1338 rows x 1 columns]

In [ ]:
features['sex'].unique()

array(['female', 'male'], dtype=object)

In [ ]:
features['sex'].value_counts()

male      676
female    662
Name: sex, dtype: int64

In [ ]:
features['smoker'].unique()

array(['yes', 'no'], dtype=object)

In [ ]:
features['smoker'].value_counts()

no     1064
yes     274
Name: smoker, dtype: int64

In [ ]:
features['region'].unique()

array(['southwest', 'southeast', 'northwest', 'northeast'], dtype=object)

In [ ]:
features['region'].value_counts()

southeast    364
southwest    325
northwest    325
northeast    324
Name: region, dtype: int64

**Encoding Categorical Data**

**One Hot Encoding**

> One hot encoder is used to convert the categorical data into numerical, so that they can be used in prediction.



In [ ]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder

In [ ]:
encoder=OneHotEncoder(sparse=False)
encoded_labels = pd.DataFrame (encoder.fit_transform(features[['sex','smoker','region']]))

In [ ]:
encoded_labels.columns = encoder.get_feature_names(['sex','smoker','region'])
dataset= pd.concat([features, encoded_labels ], axis=1)

In [ ]:
dataset

age     sex  ...  region_southeast  region_southwest
0     19.0  female  ...               0.0               1.0
1     18.0    male  ...               1.0               0.0
2     28.0    male  ...               1.0               0.0
3     33.0    male  ...               0.0               0.0
4     32.0    male  ...               0.0               0.0
...    ...     ...  ...               ...               ...
1333  50.0    male  ...               0.0               0.0
1334  18.0  female  ...               0.0               0.0
1335  18.0  female  ...               1.0               0.0
1336  21.0  female  ...               0.0               1.0
1337  61.0  female  ...               0.0               0.0

[1338 rows x 14 columns]

In [ ]:
dataset.columns

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'sex_female',
       'sex_male', 'smoker_no', 'smoker_yes', 'region_northeast',
       'region_northwest', 'region_southeast', 'region_southwest'],
      dtype='object')

In [ ]:
new_features = dataset[['age', 'sex_female', 'sex_male',
       'smoker_no', 'smoker_yes', 'bmi','children','region_northeast','region_northwest','region_southeast','region_southwest']]

**Splitting the dataset**

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:

X_train,X_test,y_train,y_test=train_test_split(new_features,target,test_size=0.3,random_state=0)
# random_state = 0 is select to get the same result

In [ ]:
print(X_train.shape)
print(X_test.shape)

(936, 11)
(402, 11)


In [ ]:
X_train

age  sex_female  ...  region_southeast  region_southwest
1163  18.0         1.0  ...               0.0               0.0
196   39.0         1.0  ...               0.0               1.0
438   52.0         1.0  ...               1.0               0.0
183   44.0         1.0  ...               0.0               0.0
1298  33.0         0.0  ...               0.0               0.0
...    ...         ...  ...               ...               ...
763   27.0         0.0  ...               0.0               0.0
835   42.0         0.0  ...               1.0               0.0
1216  40.0         0.0  ...               1.0               0.0
559   19.0         0.0  ...               0.0               0.0
684   33.0         1.0  ...               0.0               1.0

[936 rows x 11 columns]

**Fitting Multiple Linear Regression Model to Training Set**


> Different types of Regression Algorithm



1. Linear Regression (Univariate or Multivariate)
2. Support Vector Regression
3. Decision Tree Regression
4. Random Forest Regressrion

**Multivariate Linear Regression**

In [ ]:

from sklearn.linear_model import LinearRegression

regressor = LinearRegression()
regressor.fit(X_train,y_train)

# Predicting the Test Set Results
y_pred = regressor.predict(X_test)

In [ ]:
y_test

charges
578    9724.53000
610    8547.69130
569   45702.02235
1034  12950.07120
198    9644.25250
...           ...
1261   3277.16100
494   17942.10600
97    10226.28420
418   14418.28040
920   13451.12200

[402 rows x 1 columns]

In [ ]:
y_pred

array([[ 1.12531936e+04],
       [ 9.54490709e+03],
       [ 3.78498010e+04],
       [ 1.60692697e+04],
       [ 6.73440872e+03],
       [ 4.00446268e+03],
       [ 1.60576551e+03],
       [ 1.43015300e+04],
       [ 8.89696521e+03],
       [ 7.34111088e+03],
       [ 4.67363726e+03],
       [ 1.01633582e+04],
       [ 8.89221066e+03],
       [ 3.78513354e+03],
       [ 2.74730092e+04],
       [ 1.08538364e+04],
       [ 1.10975451e+04],
       [ 6.15331614e+03],
       [ 8.05810804e+03],
       [ 2.69726437e+04],
       [ 3.33303371e+04],
       [ 1.42507758e+04],
       [ 1.18538537e+04],
       [ 3.19917201e+04],
       [ 4.08938830e+03],
       [ 9.31679198e+03],
       [ 1.14341410e+03],
       [ 9.91266828e+03],
       [ 3.75184042e+03],
       [ 1.02673370e+04],
       [ 8.90007511e+03],
       [ 3.99578709e+04],
       [ 1.57594956e+04],
       [ 1.39309106e+04],
       [ 2.46369056e+04],
       [ 5.22461192e+03],
       [ 1.26648825e+04],
       [ 3.06140223e+04],
       [ 3.3

**Evaluation Metrices**

1. Mean Absolute Error (MAE)
2. Mean Squared Error (MSE)
3. R-Squared Error

In [ ]:
from sklearn.metrics import mean_absolute_error

# MAE

mean_absolute_error(y_test, y_pred)

4011.449679327985

In [ ]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, y_pred)

33342497.82695458

In [ ]:
from sklearn.metrics import r2_score

r2_score(y_test, y_pred)

0.7909160991789904

In [ ]:
dataset

age     sex  ...  region_southeast  region_southwest
0     19.0  female  ...               0.0               1.0
1     18.0    male  ...               1.0               0.0
2     28.0    male  ...               1.0               0.0
3     33.0    male  ...               0.0               0.0
4     32.0    male  ...               0.0               0.0
...    ...     ...  ...               ...               ...
1333  50.0    male  ...               0.0               0.0
1334  18.0  female  ...               0.0               0.0
1335  18.0  female  ...               1.0               0.0
1336  21.0  female  ...               0.0               1.0
1337  61.0  female  ...               0.0               0.0

[1338 rows x 14 columns]

**Feature Scaling**


> It is necessary because sometimes a large value in the variable can dominate the small value. 



In [ ]:
from sklearn.preprocessing import StandardScaler
X_sc = StandardScaler()
y_sc = StandardScaler()
X_train_new = X_sc.fit_transform(X_train[['age', 'sex_female', 'sex_male',
       'smoker_no', 'smoker_yes', 'bmi','children','region_northeast','region_northwest','region_southeast','region_southwest']])
X_test_new = X_sc.transform(X_test[['age', 'sex_female', 'sex_male',
       'smoker_no', 'smoker_yes', 'bmi','children','region_northeast','region_northwest','region_southeast','region_southwest']])
y_train_new = y_sc.fit_transform(y_train[['charges']])
y_test_new = y_sc.transform(y_test[['charges']])

**Support Vector Regression**

In [ ]:

from sklearn.svm import SVR 

regressor = SVR(kernel = 'linear')
regressor.fit(X_train_new,y_train_new)

SVR(kernel='linear')

In [ ]:
y_pred = regressor.predict(X_test_new)

In [ ]:
r2_score(y_test_new, y_pred)

0.7619821999423445

**Decision Tree Regression**

In [ ]:

from sklearn.tree import DecisionTreeRegressor


regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X_train_new,y_train_new) 

DecisionTreeRegressor(random_state=0)

In [ ]:
y_pred = regressor.predict(X_test_new)
r2_score(y_test_new, y_pred)

0.7120277244974373

**Random Forest Regression**

In [ ]:

from sklearn.ensemble import RandomForestRegressor

regressor = RandomForestRegressor(n_estimators = 10,random_state = 0) # n estiamator is the number of decision trees
regressor.fit(X_train_new,y_train_new) 

RandomForestRegressor(n_estimators=10, random_state=0)

In [ ]:
y_pred = regressor.predict(X_test_new)
r2_score(y_test_new, y_pred)

0.8405132294688319

#Result Analysis



> The R-square error for the regression are given below:




1. Linear Regression :0.7909
2. Support Vector Regression:0.7619
3. Decision Tree Regression:0.7120
4. Random Forest Regressrion:0.8405


> 


In the dataset, Multivariate Linear Regression, Support Vector Regression, Decision Tree Regression, Random Forest Regressrion has been applied. For evaluating their performances R-square error evaluation matrix has been applied for all. From the values we can see that, R-square value for random forest regression is more close to 1 which is 0.84 than the others. So that, random forest regression gives better prediction in the dataset.